### PR TITLE
Merge prototype to dev

### PR DESCRIPTION
**Pull request by imPDA**
Starting prototype created. It has layered structure (from low to high): 
- infrastructure layer: entire infrastructure (repos from which pictures and quotes gathered) and message senders to send messages
- logic layer: business logic + initialization of DI container
- application layer: for application logic

I also decided to add simple scheduler to send messages periodically, basic logging and settings (available via DI container).

Project dockerized and can be easily deployed on VPS or built and tested locally with make commands (e.x. `make up` will run application automatically) .

Quotes gathered from [quotable.io](https://quotable.io), pictures - [Lorem Picsum](https://picsum.photos/). Requires webhook to MS Teams channel to be set up! Check `.env.example` before starting container!

### CODE REVIEW

File: app/application/main.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import signal
+from datetime import datetime, timedelta
+
+from apscheduler.events import EVENT_JOB_ERROR, JobExecutionEvent
+from apscheduler.schedulers.asyncio import AsyncIOScheduler
+
+from apscheduler.triggers.interval import IntervalTrigger
+
+from logger.base import create_logger
+from logic.send_random_message import send_random_image_and_text
```
Comment (SchwarzerWolf) [inline app/application/main.py:2]: Non-pep-compliant import order

File: app/application/main.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+
+def error_listener(event: JobExecutionEvent) -> None:
+    logging.getLogger('app').exception("Job raised an exception", exc_info=event.exception)
```
Comment (SchwarzerWolf) [inline app/application/main.py]: Inconsistent quotes

File: app/application/main.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    loop = asyncio.new_event_loop()
+
+    scheduler = AsyncIOScheduler(event_loop=loop)
+    scheduler.add_listener(error_listener, EVENT_JOB_ERROR)
+
+    trigger = IntervalTrigger(minutes=60)
+    job = scheduler.add_job(send_random_image_and_text, trigger)
+
+    def graceful_stop():
```
Comment (SchwarzerWolf) [inline app/application/main.py]: Try not to define your functions inside main code - it is considered  a bad practice

File: app/application/main.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+if __name__ == '__main__':
+    logger = create_logger('app')
+    logging.getLogger('apscheduler.executors.default').disabled = True
+
+    loop = asyncio.new_event_loop()
+
+    scheduler = AsyncIOScheduler(event_loop=loop)
+    scheduler.add_listener(error_listener, EVENT_JOB_ERROR)
+
+    trigger = IntervalTrigger(minutes=60)
```
Comment (SchwarzerWolf) [inline app/application/main.py]: I propose to move all magic numbers to constants

File: app/application/main.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+    # next run scheduled at `now() + interval` by default, but I want it
+    # fire right after the loop started, so let's adjust `next_run_time`
+    job.modify(next_run_time=datetime.now() + timedelta(seconds=1))
+
+    try:
+        logger.info('App started!')
+        loop.run_forever()
+    finally:
+        loop.close()
```
Comment (SchwarzerWolf) [inline app/application/main.py]: BTW, why not to use simple cron scheduling instead of this solution?

File: app/infra/message_senders/base.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+from abc import abstractmethod, ABC
+from dataclasses import dataclass
+
+
+@dataclass
+class BaseMessageSender(ABC):
+    @abstractmethod
+    def send(self, **kwargs) -> None:
+        raise NotImplementedError()
```
Comment (SchwarzerWolf) [inline app/infra/message_senders/base.py:9]: You don't need to raise an exception here - it is done automatically then you have @abstractmethod property. It's better to substitute it with pass

File: app/infra/message_senders/teams_webhook.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
+import json
+import logging
+
+from dataclasses import dataclass
```
Comment (SchwarzerWolf) [inline app/infra/message_senders/teams_webhook.py:3]: non-pep-compliant imports

File: app/infra/message_senders/teams_webhook.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+import requests
+
+from .base import BaseMessageSender
+
+
+def generate_card(**kwargs) -> str:
+    body = []
+
+    if kwargs.get('image'):
```
Comment (SchwarzerWolf) [inline app/infra/message_senders/teams_webhook.py:13]: inconsistent quotes style

File: app/infra/repositories/picture/s3.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
+from io import BytesIO
+from dataclasses import dataclass
+from typing import Any
+
+from .base import BasePictureRepository, Picture
+
+
+@dataclass
+class S3PictureRepository(BasePictureRepository):
```
Comment (SchwarzerWolf) [inline app/infra/repositories/picture/s3.py:9]: How should your integration with s3 bucket work?

**Review by SchwarzerWolf — COMMENTED** [review]
I do like the complexity of your solution and the fact that you used patterns in your code - this way it seems more manageable. I also enjoy that you thought about the deployment of your solution.

What I don't understand: 
- How would you use a S3 bucket as a source? What settings do you need to apply to it? 
- Why have you decided to create some complicated launch script, knowing that we are currently having an Airflow course, and with 100% probability we would move this script to Airflow? 

What you should correct: 
- MR message - it's outdated 
- NotImplementedError stuff in ABC classes 
- style: inconsistent quotes, non-pep-compliant imports. I mentioned just some of them, but there are in a whole project. Try to use black + isort + flake8 combination to create pre-commit hooks if you don't want to receive those kind of comments from me 

In general - well done!

File: app/application/main.py (hunk 3 above)
Comment (imPDA) [reply app/application/main.py]: Function extracted

File: app/application/main.py (hunk 4 above)
Comment (imPDA) [reply app/application/main.py]: -> config

File: app/application/main.py (hunk 5 above)
Comment (imPDA) [reply app/application/main.py]: It is for sure possible way, but I'd like to configure everything inside script itself, including start/stop time of tasks and period.

**Comment by imPDA** [comment]
MR message updated. I also set up ruff and pre-hooks to help with formatting and linting.

File: app/application/main.py (hunk 5 above)
Comment (imPDA) [reply app/application/main.py]: Scheduler deleted after discussion

**Review by SchwarzerWolf — APPROVED** [review]

**Pull request merged**